Tests/SwiftDocCTests/Rendering/DeclarationsRenderSectionTests.swift

```diff
@@ -151,4 +152,114 @@ class DeclarationsRenderSectionTests: XCTestCase {
         XCTAssertEqual(declarationsSection.declarations.count, 2)
         XCTAssert(declarationsSection.declarations.allSatisfy({ $0.platforms == [.iOS, .macOS] }))
     }
+
+    func testHighlightDiff() throws {
```

Comment:
I think it would be good to add more—and more complicated—tests for this. The code works really well but it'd be good to have more comprehensive testing so that we can be confident that we don't break it in the future.

This is just one test with two overloads `func myFunc(param: Int)` and `func myFunc<S>(param: S) where S : StringProtocol` but there's a lot more possible syntax that would be worth testing, especially for text tokens since there's logic to split those for better diffident. 

Here are the variations that I could think of for a single parameter type. We definitely don't need to add tests for all of them but I think we should have something that's an array, something with a dictionary, something with a tuple, something with an optional, something with a closure type, something with a variadic, and something with a generic argument.

```swift
public func doSomething(with: Int)              {}
public func doSomething(with: Int?)             {}
public func doSomething(with: [Int]?)           {}
public func doSomething(with: [Int?])           {}
public func doSomething(with: (Int, Int))       {}
public func doSomething(with: (Int?, Int))      {}
public func doSomething(with: (Int, Int?))      {}
public func doSomething(with: (Int, Int)?)      {}
public func doSomething(with: [Int: Int])       {}
public func doSomething(with: [Int: Int]?)      {}
public func doSomething(with: [Int?: Int])      {}
public func doSomething(with: [Int: Int?])      {}
public func doSomething(with: Int...)           {}
public func doSomething(with: Int?...)          {}
public func doSomething(with: Set<Int>)         {}
public func doSomething(with: Set<Int?>)        {}
public func doSomething(with: Set<Int>?)        {}
public func doSomething(with: (Int) -> ())      {}
public func doSomething(with: (Int) -> Int)     {}
public func doSomething(with: (Int?) -> Int)    {}
public func doSomething(with: (Int) -> Int?)    {}
public func doSomething(with: ((Int) -> Int)?)  {}
```

Reply:
(by the way, the diff in this test is not that "fancy". I can come think with much more complex differences between overloads)

Reply:
The existing code seems to trip up with tuples and closures, because the differencing eagerly gloms onto the first closing parenthesis it finds and treats that as a common token with the closing parenthesis for the argument list in other overloads...

<img width="469" alt="image" src="https://github.com/apple/swift-docc/assets/5217170/7ee02087-be5b-4657-b99d-d709996955d7">

It also makes the whitespace trimming fall apart for where clauses, since now the argument list parenthesis is considered a different token:

<img width="612" alt="image" src="https://github.com/apple/swift-docc/assets/5217170/8219d596-4357-4bfc-8b1a-bc985fd0779e">

I can write a test with these symbols, but the highlighting here is a bit unfortunate. However, to fix it "properly" would require introducing the complete symbol information into the differencing algorithm somehow, so that the entire argument type could be considered a distinct "token" and the correct parenthesis (and comma, in case of multiple arguments) could be counted as a common token. If we decide that we want to improve this, i'd like to defer that to after this PR lands so that we can get a "good-enough" implementation landed that we can iterate on.

Reply:
The highlighting is great. I don't think the few cases where it could be slightly better should hold back this PR.

If anything we could add a comment in the tests for the highlights that could be slightly better as extra information for anyone who wants to iterate on this in the future. 

In other words: the implementation looks great and we don't need to add many new tests but I think we should add a handful (maybe one with a tuple, one with a closure, and one with an array/dictionary and then fit an optional and a generic value in one of those 3 cases). How does that sound?

Reply:
I've updated my tests to test the following overload groups:

```swift
public func overload1(param: Int) {}
public func overload1(param: Int?) {}
public func overload1(param: [Int]) {}
public func overload1(param: [Int]?) {}
public func overload1(param: Set<Int>) {}
public func overload1(param: [Int: Int]) {}

public func overload2(p1: Int, p2: Int) {}
public func overload2(p1: (Int, Int), p2: Int) {}
public func overload2(p1: Int, p2: (Int, Int)) {}
public func overload2(p1: (Int) -> (), p2: Int) {}
public func overload2(p1: (Int) -> Int, p2: Int) {}
public func overload2(p1: (Int) -> Int?, p2: Int) {}
public func overload2(p1: ((Int) -> Int)?, p2: Int) {}

public func overload3(_ p: [Int: Int]) {}
public func overload3<T: Hashable>(_ p: [T: T]) {}
public func overload3<K: Hashable, V>(_ p: [K: V]) {}
```

I feel like this both tests the features we want (type decorators getting highlighted, whitespace trimmed off highlighted tokens, splitting `>(` tokens) and also adds the known edge case issues (parentheses and commas throwing off the diff).

I also added a convenience wrapper in the test code so that i could more concisely test that certain spans of declarations were being highlighted as expected. I'm not 100% sure that this is completely useful, but it helped when rewriting the tests to work with six or seven overloads at a time. 😅

Reply:
After some out-of-band discussion, i've rewritten the test wrapper to render plain-text comparison strings instead of using the string fragments i originally used.